File: 13 Classification and Regression Trees and Random Forests/Decision Tree.py
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
from sklearn.tree import DecisionTreeClassifier, plot_tree, export_text
from sklearn.metrics import confusion_matrix

#%%
df = pd.read_csv("../Datasets/fin-ratio.csv")
X = df.drop(columns="HSI")
y = df["HSI"]
ctree = DecisionTreeClassifier(ccp_alpha=0.01)
ctree.fit(X, y)
print(export_text(ctree, feature_names=list(X.columns), 
                  show_weights=True))

fig, ax = plt.subplots(1, 1, figsize=(20, 15))
plot_tree(ctree, feature_names=X.columns, filled=True)
fig.tight_layout()
fig.savefig("../Picture/Classification tree fin-ratio.png", dpi=200)

#%%
fig, ax = plt.subplots(1, 1, figsize=(10, 9))
scatter = ax.scatter(y, X.ln_MV, c=y, alpha=0.7,
                     cmap=ListedColormap(["blue", "red"]))
ax.legend(handles=scatter.legend_elements()[0], loc="upper left", 
          labels=["HSI", "Non-HSI"], fontsize=15)
ax.axhline(y=9.478)
ax.set_xlabel("HSI", fontsize=15)
ax.set_ylabel("ln MV", fontsize=15)
fig.tight_layout()
fig.savefig("../Picture/ln MV vs HSI.png", dpi=400)

#%%
y_hat = ctree.predict(X)
print(confusion_matrix(y_hat, y))

#%%
from sklearn.datasets import load_iris

iris = load_iris()
X, y = iris["data"], iris['target']
ctree = DecisionTreeClassifier(ccp_alpha=0.01, random_state=4012)
ctree.fit(X, y)

fig, ax = plt.subplots(1, 1, figsize=(20, 15))
plot_tree(ctree, feature_names=iris['feature_names'], filled=True)
fig.tight_layout()
fig.savefig("../Picture/Classification tree Iris flower.png", dpi=200)

print(export_text(ctree, feature_names=iris['feature_names'], 
                  show_weights=True))

#%%
fig, ax = plt.subplots(1, 1, figsize=(10, 9))
scatter = ax.scatter(X[:,2], X[:,3], c=y, alpha=0.7,
                     cmap=ListedColormap(["red", "blue", "green"]))
ax.legend(handles=scatter.legend_elements()[0], loc="upper left", 
          labels=list(iris['target_names']), fontsize=15)
plt.axhline(y=1.75)
plt.axvline(x=2.45)
plt.axhline(y=1.65)
plt.axvline(x=4.95)
ax.set_xlabel("Petal length", fontsize=15)
ax.set_ylabel("Petal width", fontsize=15)
fig.tight_layout()
fig.savefig("../Picture/Petal width vs Petal length.png", dpi=400)

#%%
y_hat = ctree.predict(X)
print(confusion_matrix(y_hat, y))

#%%
# Regression Tree

import pandas as pd
from sklearn.tree import DecisionTreeRegressor, plot_tree, export_text
import matplotlib.pyplot as plt
from matplotlib import collections  as mc

# Prepare the dataset
df = pd.read_csv("../Datasets/Medicalpremium.csv")
X = df[['Age', 'Weight']]
y = df['PremiumPrice']

# Fit and plot the regression tree model
rtree = DecisionTreeRegressor(ccp_alpha=0.01, max_depth=3)
rtree.fit(X, y)
print(export_text(rtree, feature_names=list(X.columns), 
                  show_weights=True))

plt.figure(figsize=(12, 8))
plot_tree(rtree, feature_names=list(X.columns), filled=True)

plt.tight_layout()
plt.savefig("../Picture/Medical Insurance RTree.png", dpi=200)

# Plot dataset with segments and text annotations
fig, ax = plt.subplots(figsize=(8, 6))
s = ax.scatter(df['Age'], df['Weight'], c=y, cmap='rainbow')

# Add vertical lines and text annotations
ax.axvline(x=29.5, color='black', linewidth=2)
ax.axvline(x=46.5, color='black', linewidth=2)
ax.axvline(x=38.5, color='black', linewidth=2)
lines = [[(0, 119), (29.5, 119)], 
         [(24.5, 0), (24.5, 119)], [(23.0, 119), (23.0, 140)],
         [(46.5, 94.5), (70, 94.5)]]
lc = mc.LineCollection(lines, colors="black", linewidths=2)
ax.add_collection(lc)
plt.text(18.5, 127, "R1", fontsize=12)
plt.text(25.5, 127, "R2", fontsize=12)
plt.text(19.5, 85, "R3", fontsize=12)
plt.text(26, 85, "R4", fontsize=12)
plt.text(33, 90, "R5", fontsize=12)
plt.text(41.5, 90, "R6", fontsize=12)
plt.text(56, 115, "R7", fontsize=12)
plt.text(56, 71, "R8", fontsize=12)

plt.xlabel('Age')
plt.ylabel('Weight')
fig.colorbar(s, label='Premium Price')

fig.tight_layout()
fig.savefig("../Picture/Medical Insurance Scatter.png", dpi=200)

#%%
from sklearn.ensemble import RandomForestClassifier

df = pd.read_csv("../Datasets/fin-ratio.csv")
X = df.drop(columns="HSI")
y = df["HSI"]
rf_clf = RandomForestClassifier(max_features=2, random_state=4002)
rf_clf.fit(X, y)
y_hat = rf_clf.predict(X)
print(confusion_matrix(y_hat, y))

#%%
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

df = pd.read_csv("../Datasets/credit default.csv")
X = df.drop(columns=["default payment next month"])
y = df["default payment next month"]

(X_train, X_test, y_train, 
 y_test) = train_test_split(X, y, train_size=0.8, random_state=4012)
ctree = DecisionTreeClassifier(ccp_alpha=0.01, random_state=4012)
ctree.fit(X_train, y_train)
y_hat_dt = ctree.predict(X_test)
print(confusion_matrix(y_hat_dt, y_test))
print(classification_report(y_test, y_hat_dt))

fig, ax = plt.subplots(1, 1, figsize=(20, 15))
plot_tree(ctree, feature_names=X.columns, filled=True)
fig.tight_layout()
fig.savefig("../Picture/Credit Default DT.png", dpi=200)

rf_clf = RandomForestClassifier(random_state=4012)
rf_clf.fit(X_train, y_train)
y_hat_rf = rf_clf.predict(X_test)
print(confusion_matrix(y_hat_rf, y_test))
print(classification_report(y_test, y_hat_rf))

File: 13 Classification and Regression Trees and Random Forests/Decision Tree.R
#Set directory: Run this on source instead of Console!!
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

################################################################################
library(rpart)                          # load rpart library
library(rpart.plot)                     # plot rpart object

################################################################################
### Financial ratio ###
################################################################################
df <- read.csv("../Datasets/fin-ratio.csv")         # read in data in csv format
ctree <- rpart(HSI~., data=df, method="class")
print(ctree)                            # print detailed information
rpart.rules(ctree, nn=TRUE)             # print classification rules
rpart.plot(ctree, extra=1, cex=1.5, digits=4, nn=TRUE) # plot ctree

################################################################################
plot(df$HSI, df$ln_MV, pch=21, bg=c("red","blue")[df$HSI+1])
legend(0, 14, legend=c("Non-HSI", "HSI"), pch=16, col=c("red","blue"))
abline(h=9.478)		# add a horizontal line at y=9.478

################################################################################
prob <- predict(ctree)      # 2 columns of probabilities for 0 or 1
y_hat <- colnames(prob)[max.col(prob)]
table(y_hat, df$HSI)		                # confusion matrix

################################################################################
### IRIS flower ###
################################################################################
data("iris")                # load the built-in iris flower dataset
df <- iris
ctree <- rpart(Species~., data=df, method="class")
print(ctree)                            # print detailed information
rpart.rules(ctree, nn=TRUE)             # print classification rules
rpart.plot(ctree, extra=1, cex=1.5, digits=4, nn=TRUE) # plot ctree

################################################################################
# plot Petal.Width versus Petal.Length with different color for each species
plot(df$Petal.Length, df$Petal.Width, pch=21, 
     bg=c("red","blue","green")[df$Species])
legend(1, 2.5, legend=unique(df$Species), pch=16, 
       col=c("red", "blue", "green"))
abline(h=1.75)		                      # add a horizontal line
abline(v=2.45)		                      # add a vertical line

################################################################################
prob <- predict(ctree)  # 3 columns of probabilities for 3 species
y_hat <- colnames(prob)[max.col(prob)]
table(y_hat, df$Species)                # confusion matrix

################################################################################
### Regression Tree ###
################################################################################
library(rpart)                          # load rpart library
library(rpart.plot)                     # plot rpart object

# Prepare the dataset
df <- read.csv("../Datasets/Medicalpremium.csv")

# Fit and plot the regression tree model
rtree <- rpart(PremiumPrice~Age+Weight, data=df, method="anova")
rpart.rules(rtree, nn=TRUE)
rpart.plot(rtree, extra=0, cex=1, digits=2, type=0, nn=TRUE)

# Plot dataset with segments and text annotations
layout(t(1:2), widths=c(6,1))
par(mar=c(4,4,1,1))
n_col <- 20
y <- df$PremiumPrice
y_scale <- round((y - min(y))/diff(range(y))*(n_col-1)) + 1
plot(df$Age, df$Weight, pch=20, xlab="Age", ylab="Weight", 
     col=rainbow(n_col)[y_scale])
abline(v=30, lwd=2)
abline(v=47, lwd=2)
segments(47, 70, 70, 70, lwd=2)
segments(47, 95, 70, 95, lwd=2)
text(22, 90, "R1", cex=1.5)
text(38, 90, "R2", cex=1.5)
text(58, 60, "R3", cex=1.5)
text(58, 82, "R4", cex=1.5)
text(58, 110, "R5", cex=1.5)
image(y=1:n_col, z=t(1:n_col), col=rainbow(n_col), axes=FALSE, 
      main="Premium", cex.main=.8, ylab="")

################################################################################
### Random Forest ###
################################################################################
library(randomForest)

set.seed(4002)
df <- read.csv("../Datasets/fin-ratio.csv")         # read in data in csv format
df$HSI <- as.factor(df$HSI) # change label into factor for classification
rf_clf <- randomForest(HSI~., data=df, ntree=10, mtry=2, importance=TRUE)
y_hat <- predict(rf_clf, newdata=df)
table(y_hat, df$HSI)

################################################################################
library("caret")                        # confusionMatrix

set.seed(4002)				                  # set random seed
df <- read.csv("../Datasets/credit default.csv")    # read in data in csv format
df$default.payment.next.month <- as.factor(
  df$default.payment.next.month
)                       # change label into factor for classification

train_idx <- sample(1:nrow(df), size=floor(nrow(df)*0.8))
df_train <- df[train_idx,]		          # training dataset
df_test <- df[-train_idx,]		          # testing dataset

ctree <- rpart(default.payment.next.month~., data=df_train, 
               method="class")
rpart.plot(ctree, extra=1, cex=1.5, digits=4, nn=TRUE) # plot ctree
prob <- predict(ctree, newdata=df_test)
y_hat_dt <- colnames(prob)[max.col(prob)]

# confusionMatrix(y_test, y_true, ...)
dt_result <- confusionMatrix(as.factor(y_hat_dt), 
                             df_test$default.payment.next.month, 
                             mode="prec_recall", positive="1")
dt_result$table         # Confusion matrix
dt_result$byClass[c("Precision", "Recall")]

rf_clf <- randomForest(default.payment.next.month~., data=df_train, 
                       ntree=10, importance=TRUE)
y_hat_rf <- predict(rf_clf, newdata=df_test)
rf_result <- confusionMatrix(as.factor(y_hat_rf), 
                             df_test$default.payment.next.month, 
                             mode="prec_recall", positive="1")
rf_result$table            # Confusion matrix
rf_result$byClass[c("Precision", "Recall")]